src/main/java/com/warnermedia/kplserver/KinesisEventPublisher.java

```diff
@@ -0,0 +1,106 @@
+package com.warnermedia.kplserver;
+
+import com.amazonaws.services.kinesis.producer.KinesisProducer;
+import com.amazonaws.services.kinesis.producer.KinesisProducerConfiguration;
+import com.amazonaws.services.kinesis.producer.UserRecord;
+import com.amazonaws.services.kinesis.producer.UserRecordFailedException;
+import com.amazonaws.services.kinesis.producer.UserRecordResult;
+import com.amazonaws.services.sqs.AmazonSQS;
+import com.amazonaws.services.sqs.AmazonSQSClientBuilder;
+import com.amazonaws.services.sqs.model.SendMessageRequest;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import java.math.BigInteger;
+import java.nio.ByteBuffer;
+import java.util.Random;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.stream.Collectors;
+
+public class KinesisEventPublisher {
+  private final String stream;
+  private static final Log log = LogFactory.getLog(
+    KinesisEventPublisher.class);
+  final ExecutorService callbackThreadPool = Executors.newCachedThreadPool();
+  private final AmazonSQS sqs;
+  private final KinesisProducer kinesis;
+  private static final String queueUrl = System.getenv("DLQ_URL");
+
+  public KinesisEventPublisher(String stream, String region) {
+    this.stream = stream;
+    kinesis = new KinesisProducer(new KinesisProducerConfiguration()
+      .setRegion(region));
+    sqs = AmazonSQSClientBuilder.standard().withRegion(region).build();
+  }
+
+  public void runOnce(String line) throws Exception {
+    // add new line so that downstream systems have an easier time parsing
+    String finalLine = line + "\n";
+
+    ByteBuffer data = ByteBuffer.wrap(line.getBytes(java.nio.charset.StandardCharsets.UTF_8));
+
+    //This is a measure of the backpressure in the system, which should be checked before putting more records,
+    //to avoid exhausting system resources.
+    while (kinesis.getOutstandingRecordsCount() > 1e4) {
+      log.info("Too many outstanding records pending in the queue. Waiting for a second.");
+      Thread.sleep(500);
+    }
```

Comment:
Was it recommended somewhere to do this? I'm just a little apprehensive because of the round-trip (presumably) for `getOutstandingRecordsCount()`.

Reply:
Yes. This one https://aws.amazon.com/blogs/big-data/implementing-efficient-and-reliable-producers-with-the-amazon-kinesis-producer-library/ .

Reply:
It does not explain why though. I think the reason is that we have a limit in kinesis stream on number of messages we can write at every second per shard. If we cross that limit, all new messages will start dropping by way of backpressure.